### ISSUE TITLE
[false positive] [clang-tidy] [clang-analyzer-core.uninitialized.Assign]

### ISSUE THREAD

**Issue by RalphS69** [llvm/llvm-project#54375]
With clang-tidy (LLVM version 13.0.0). , this code snipped
```
#include <assert.h>
#include <stdint.h>

#define SIZE 1

void func( int16_t len1, int16_t len2 )
{
    int16_t i;
    float t[SIZE];

    assert( len1 <= SIZE );
    assert( len2 <= len1 );
            
    for ( i = 0; i < len1; i++ )
    {
        t[i] = 1;
    }
            
    for ( i = 0; i < len2; i++ )
    {
        float s = t[0];
        (void) s;
    }
}
```
 causes 
```
1 warning generated.
warning: Assigned value is garbage or undefined [clang-analyzer-core.uninitialized.Assign]
        float s = t[0];
        ^         ~~~~
note: Assuming 'len1' is <= 1
    assert( len1 <= SIZE );
            ^
Taking true branch
    assert( len1 <= SIZE );
    ^
note: Assuming 'len2' is <= 'len1'
    assert( len2 <= len1 );
            ^
note: Taking true branch
    assert( len2 <= len1 );
    ^
note: Assuming 'i' is >= 'len1'
    for ( i = 0; i < len1; i++ )
                 ^~~~~~~~
note: Loop condition is false. Execution continues on line 19
    for ( i = 0; i < len1; i++ )
    ^
note: Assuming 'i' is < 'len2'
    for ( i = 0; i < len2; i++ )
                 ^~~~~~~~
note: Loop condition is true.  Entering loop body
    for ( i = 0; i < len2; i++ )
    ^
note: Assigned value is garbage or undefined
        float s = t[0];
        ^         ~~~~
```

- The assumption `'i' is >= 'len1'` leads with `i = 0` to **len1 <= 0**.
- with `assert( len2 <= len1 )` we assure **len2 <= 0**.
- How can then `'i' is < 'len2'` be assumed? With` i = 0` this leads to **len2 > 0**, which is a contradiction to the first two statements.

In other words: How can the second loop body be ever entered, it the first loop body was not entered, if both loops start with 0, if the index in both loops is incremented by one, and if the number of iterations of the second loop is not larger than the number of iterations of the first loop?

**Comment by haoNoQ** [comment]
This is an example of #39492. Currently we have
```
reg_$0<int16_t len1> { [-32768, 0] }
reg_$1<int16_t len2> { [1, 32767] }
(reg_$1<int16_t len2>) <= (reg_$0<int16_t len1>) { [-2147483648, -1], [1, 2147483647] }
```
so it should be somewhat straightforward to collapse  `len2 <= len1` to `false`. Of course we need to be careful about promotion rules but here both values are signed and short so promotion to `int` doesn't change much. There's also a liveness bug a-la #53176 here which causes us to forget about some of these constraints too early so we never see them together unless `len1` is used later.

cc @martong

**Comment by martong** [comment]
The crux of this false positive report is that we cannot handle transitivity of the `<=` operator. This is because we have a mechanism that reevaluates all existing constraints if a symbol is constrained with a constant, however, we don't do that if the symbol is constrained with a range. Here is the test that demonstrates this.
```
// RUN: %clang_analyze_cc1 %s \
// RUN:   -analyzer-checker=core \
// RUN:   -analyzer-checker=debug.ExprInspection \
// RUN:   -analyzer-config eagerly-assume=false \
// RUN:   -verify

void clang_analyzer_eval(int);

extern void abort() __attribute__((__noreturn__));
#define assert(expr) ((expr) ? (void)(0) : abort())

void func(int len1, int len2) {
  assert(len1 <= 1);
  assert(len2 <= len1);

  if (0 < len1) { // Simplification kicks because len1 is constrained to a constant.
    clang_analyzer_eval(len1 == 1);    // expected-warning{{TRUE}};
    clang_analyzer_eval(len2 <= 1);    // expected-warning{{TRUE}};
  } else {        // No simplification, len1 is constrained with a range.
    clang_analyzer_eval(len1 <= 0);    // expected-warning{{TRUE}};
    clang_analyzer_eval(len2 <= 0);    // expected-warning{{UNKNOWN}}; should be TRUE
  }

  (void)(len1 + len2);
}
```
To support transitivity of LT, GT, LE, GE operators, we should do something similar to the existing simplification mechanism.
This would require something similar to `EquivalenceClass::simplify` and a similar visitor that we have in `SimpleSValBuilder::simplifySValOnce`.  This would certainly increase the strength of our range based constraint solver.
FYI @steakhal
